pkg/network/http/ebpf_gotls.go

```diff
@@ -158,77 +252,122 @@ func supportedArch(arch string) bool {
 	return arch == string(bininspect.GoArchX86_64)
 }
 
-func (p *GoTLSProgram) handleNewBinary(binPath string) {
-	f, err := os.Open(binPath)
+func (p *GoTLSProgram) handleProcessStart(pid pid) {
+	exePath := filepath.Join(p.procRoot, strconv.FormatUint(uint64(pid), 10), "exe")
+	binPath, err := os.Readlink(exePath)
 	if err != nil {
-		log.Errorf("Could not open file %q due to %s", binPath, err)
+		log.Debugf(" could not read binary path for pid %d: %s", pid, err)
 		return
 	}
-	defer f.Close()
-	elfFile, err := elf.NewFile(f)
-	if err != nil {
-		log.Errorf("File %q could not be parsed as elf: %s", binPath, err)
+
+	var stat syscall.Stat_t
+	if err = syscall.Stat(binPath, &stat); err != nil {
+		log.Errorf("could not stat binary path %s: %s", binPath, err)
 		return
 	}
 
-	result, err := bininspect.InspectNewProcessBinary(elfFile, functionsConfig, structFieldsLookupFunctions)
-	if err != nil {
-		log.Errorf("Failed inspecting binary %q: %s", binPath, err)
+	hookedBin := p.analysisCache.Get(stat.Ino)
+	if hookedBin == nil {
+		f, err := os.Open(binPath)
+		if err != nil {
+			log.Errorf("could not open file %s, %s", binPath, err)
+			return
+		}
+		defer f.Close()
+
+		elfFile, err := elf.NewFile(f)
+		if err != nil {
+			log.Errorf("file %s could not be parsed as an ELF file: %s", binPath, err)
+			return
+		}
+
+		inspectionResult, err := bininspect.InspectNewProcessBinary(elfFile, functionsConfig, structFieldsLookupFunctions)
+		if err != nil {
+			if !errors.Is(err, binversion.ErrNotGoExe) {
+				log.Errorf("error reading exe: %s", err)
+			}
+			return
+		}
+
+		if err = p.addInspectionResultToMap(inspectionResult, stat.Ino); err != nil {
+			log.Error(err)
+			return
+		}
+
+		probeIDs, err := p.attachHooks(inspectionResult, binPath)
+		if err != nil {
+			log.Errorf("error while attaching hooks: %s", err)
+			p.removeInspectionResultFromMap(stat.Ino)
+			return
+		}
+		log.Debugf("attached hooks on %s (%d)", binPath, stat.Ino)
+		hookedBin = p.analysisCache.Set(stat.Ino, &hookedBinary{probeIDs, make(runningProcessesSet)}, ttlcache.NoTTL)
```

Comment:
that's a long logic inside, I think it can be better to extract it another function
Leaving that up to you